gui/widgets/groupedlist.cpp

```diff
@@ -0,0 +1,427 @@
+	const frac_t threshold = intToFrac(15) / 16;
+
+	if ((frac_t)(entriesPerPage & FRAC_LO_MASK) >= threshold)
+		entriesPerPage += FRAC_ONE;
+
+	_entriesPerPage = fracToInt(entriesPerPage);
+	assert(_entriesPerPage > 0);
+
+	if (_scrollBar) {
+		_scrollBar->resize(_w - _scrollBarWidth, 0, _scrollBarWidth, _h, false);
+		scrollBarRecalc();
+		scrollToCurrent();
+	}
+}
+
+void GroupedListWidget::toggleGroup(int groupID) {
+	// Shrink group if it is expanded
+	if (_groupExpanded[groupID]) {
+		_groupExpanded[groupID] = false;
+		sortGroups();
+	} else {
+	// Expand group if it is shrunk
+		_groupExpanded[groupID] = true;
+		sortGroups();
+	}
+}
+
+void GroupedListWidget::drawWidget() {
+	int i, pos, len = _list.size();
+	Common::U32String buffer;
+
+	// Draw a thin frame around the list.
+	g_gui.theme()->drawWidgetBackground(Common::Rect(_x, _y, _x + _w, _y + _h),
+	                                    ThemeEngine::kWidgetBackgroundBorder);
+
+	// Draw the list items
+	for (i = 0, pos = _currentPos; i < _entriesPerPage && pos < len; i++, pos++) {
+		const int y = _y + _topPadding + kLineHeight * i;
+		const int fontHeight = kLineHeight;
+		ThemeEngine::TextInversionState inverted = ThemeEngine::kTextInversionNone;
+
+		// Draw the selected item inverted, on a highlighted background.
+		if (_selectedItem == pos)
+			inverted = _inversion;
+
+		Common::Rect r(getEditRect());
+		int pad = _leftPadding;
+		int rtlPad = (_x + r.left + _leftPadding) - (_x + _hlLeftPadding);
+
+		if (isGroupHeader(_listIndex[pos])) {
+			int groupID = indexToGroupID(_listIndex[pos]);
+			r.left += fontHeight - 2 + _leftPadding;
+			g_gui.theme()->drawFoldIndicator(Common::Rect(_x + _hlLeftPadding + _leftPadding, y, _x + fontHeight - 2 + _leftPadding, y + fontHeight - 2), _groupExpanded[groupID]);
+			pad = 0;
+		}
+
+		// If in numbering mode & not in RTL based GUI, we first print a number prefix
+		if (_numberingMode != kListNumberingOff && g_gui.useRTL() == false) {
+			buffer = Common::String::format("%2d. ", (pos + _numberingMode));
+			g_gui.theme()->drawText(Common::Rect(_x + _hlLeftPadding, y, _x + r.left + _leftPadding, y + fontHeight - 2),
+									buffer, _state, _drawAlign, inverted, _leftPadding, true);
+			pad = 0;
+		}
+
+		ThemeEngine::FontColor color = ThemeEngine::kFontColorNormal;
+
+		if (!_listColors.empty()) {
+			if (_filter.empty() || _selectedItem == -1)
+				color = _listColors[pos];
+			else
+				color = _listColors[_listIndex[pos]];
+		}
+
+		Common::Rect r1(_x + r.left, y, _x + r.right, y + fontHeight - 2);
+
+		if (g_gui.useRTL()) {
+			if (_scrollBar->isVisible()) {
+				r1.translate(_scrollBarWidth, 0);
+			}
+
+			if (_numberingMode != kListNumberingOff) {
+				r1.translate(-rtlPad, 0);
+			}
+		}
+
+		if (_selectedItem == pos && _editMode) {
+			buffer = _editString;
+			color = _editColor;
+			adjustOffset();
+		} else {
+			buffer = _list[pos];
+		}
+		if (isGroupHeader(_listIndex[pos]))
```

Comment:
Consider assigning the style-choice to a variable, so that we can use a single call to drawText.

Reply:
Fixed in 1d8a83e.